### ISSUE TITLE
LineLengthCheck should check for import and package statements

### ISSUE THREAD

**Issue by pablo-ml** [checkstyle/checkstyle#9999]
This is related to the fix originated in this bug: https://github.com/checkstyle/checkstyle/issues/2433

On the first comment, I found this:

"""
"imports" are not possible to line-wrap so LineLength should skip them from validation. Special option is not required for this, Javadoc/Xdoc should be updated.
"""

AFAIK, the JLS supports splitting the import:

````
SingleTypeImportDeclaration:
   import TypeName ;

TypeName:
  TypeIdentifier
  PackageOrTypeName . TypeIdentifier
````
So, an import production is of the form:

`import identifier . identifier . identifier .... ;`

ie: a sequence of tokens. AFAIK, whitespaces are accepted between tokens, so technically, you can write an import like this:

````
import my.package.name.
   MyClass;
````

I think that the LineLengthCheck check must be either modified back to the original behaviour, or at least add an option to check for line lengths on import and package statements.

```
bash
~/.../mp/aa$ javac A.java 
~/.../mp/aa$ 

~/.../mp/aa$ cat check.xml 
<!DOCTYPE module PUBLIC
    "-//Puppy Crawl//DTD Check Configuration 1.2//EN"
    "http://www.puppycrawl.com/dtds/configuration_1_2.dtd">

<module name="Checker">
  <module name="LineLength">
    <property name="max" value="20"/>
  </module>
</module>

~/.../mp/aa$ cat A.java 
import java . util . List;

class A {
};

~/.../mp/aa$ RUN_LOCALE="-Duser.language=en -Duser.country=US"
~/.../mp/aa$ java -jar ~/Downloads/checkstyle-8.42-all.jar -c check.xml A.java
Starting audit...
Audit done.
```
---

I expect the check to fail:

````
~/.../mp/aa$ java -jar ~/Downloads/checkstyle-8.42-all.jar -c check.xml A.java
Starting audit...
[ERROR] /home/pagrana/projects/mp/aa/A.java:1: Line is longer than 20 characters (found 26). [LineLength]
Audit done.
Checkstyle ends with 1 errors
````

---

**Comment by nmancus1** [comment]
A temporary workaround could be:

```bash
➜  src cat Test.java                                            
import com.mycompanywithreallyreallyreallysuperlongnamethatviolateslinelength.somepackage.someotherpackage.Thing;
public class Test { }

➜  src cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
        <module name="Regexp">
            <property name="format" value="^.{100,100}"/>
            <property name="illegalPattern" value="true"/>
            <message key="illegal.regexp"
                     value="Line length is longer than 100 characters."
            />
        </module>
    </module>
</module>

➜  src java -jar checkstyle-8.42-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/Desktop/Tester/src/Test.java:1: Line length is longer than 100 characters. [Regexp]
Audit done.
Checkstyle ends with 1 errors.

```

Line wrapped imports do compile:

```bash

➜  src cat Test.java
import java.
        util.
        HashMap;

public class Test { }

➜  src javac Test.java

➜  src java -version
openjdk version "1.8.0_282"
OpenJDK Runtime Environment (build 1.8.0_282-b08)
OpenJDK 64-Bit Server VM (build 25.282-b08, mixed mode)

```

**Comment by pablo-ml** [comment]
Thanks a lot for the workaround!. I completely forgot about it.

I still want to point out that the rationale on the bug https://github.com/checkstyle/checkstyle/issues/2433 was not accurate, given that the JLS supports splitting import lines. I think that the correct behaviour of the LineLength rule it to also apply to import and package names, and probably add a new flag to optionally exclude import and package statements from the check to make everybody happy :).

Thanks again!

**Comment by romani** [comment]
confirmed that it is allowed:
```
✔ ~/java/github/romani/checkstyle [haanhvu/issue#9889 L|✚ 1⚑ 1] 
 $ git diff
diff --git a/src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java b/src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java
index a0b779b..8e05b16 100644
--- a/src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java
@@ -17,7 +17,8 @@
 // Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 ////////////////////////////////////////////////////////////////////////////////
 
-package com.puppycrawl.tools.checkstyle.utils;
+package com.puppycrawl.tools.
+        checkstyle.utils;
 
 import java.util.ArrayList;
 import java.util.List;
@@ -33,7 +34,8 @@ import com.puppycrawl.tools.checkstyle.checks.javadoc.InvalidJavadocTag;
 import com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocTag;
 import com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocTagInfo;
 import com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocTags;
-import com.puppycrawl.tools.checkstyle.checks.javadoc.utils.BlockTagUtil;
+import com.puppycrawl.tools.checkstyle.checks
+        .javadoc.utils.BlockTagUtil;
 import com.puppycrawl.tools.checkstyle.checks.javadoc.utils.InlineTagUtil;
 import com.puppycrawl.tools.checkstyle.checks.javadoc.utils.TagInfo;
 
✔ ~/java/github/romani/checkstyle [haanhvu/issue#9889 L|✚ 1⚑ 1] 
$ mvn clean compile
...
[INFO] BUILD SUCCESS
```

@rnveach , original issue was created based on your PR experience ... 
please recheck this in Eclipse, as Eclipse compiler might have problems with it.
If so, we need to make it configurable.

**Comment by rnveach** [comment]
@romani Current version of Eclipse I am using has no issues. I don't remember my Eclipse at the time being having any issues.
Going by PR history at https://github.com/checkstyle/checkstyle/pull/2429 this was related to limiting all lines to 100 characters and I was asking where to break imports to remove suppressions. That was what prompted the issue to be created.

**Comment by romani** [comment]
I do not remember rationale at that time.
But it is clear that there are a lot of people who do not care about package and import length, as nobody read such lines usually.
Imports done by IDE and they do not wrap.

So we have a request to keep validation of this, ok let's have an option for such people

**Comment by rnveach** [comment]
I will mark this as approved then to add those 2 properties.

**Comment by Vyom-Yadav** [comment]
I am on it.

**Comment by Vyom-Yadav** [comment]
@romani @nick-mancuso This was addressed by #11398. Please close this issue.

**Comment by nick-mancuso** [comment]
@Vyom-Yadav please show proof via CLI output using exact example from issue description.

**Comment by Vyom-Yadav** [comment]
> @Vyom-Yadav please show proof via CLI output using exact example from issue description.

`config.xml`
```xml
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.2//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_2.dtd">

<module name="Checker">
  <module name="LineLength">
    <property name="max" value="10"/>
    <property name="ignorePattern" value="^$"/>
  </module>
</module>
```
`Test.java`
```java
package TestP;

import java . util . List;

class A {
}
```
```
$ java -jar checkstyle-10.1-all.jar -c config.xml  Test.java                                                                                                         
[ERROR] /var/TestP/Test.java:1: Line is longer than 10 characters (found 14). [LineLength]
[ERROR] /var/TestP/Test.java:3: Line is longer than 10 characters (found 26). [LineLength]
Audit done.
Checkstyle ends with 2 errors.
```

**Comment by romani** [comment]
Check does ignorance of packages and imports https://checkstyle.org/config_sizes.html#LineLength_Properties

> Specify pattern for lines to ignore. | Pattern | "^(package\|import) .*"

Default value is set like this to keep compatibility. If user wants to be more strict, he needs to update config  with `<property name="ignorePattern" value="^$"/>`, see example above.
